[lm-44] iP

ty-3401 opened the pull request:
<h1>Cook</h1>

> "Time you enjoy wasting is not wasted time." - Marthe Troly-Curtin ([sauce](https://quoteinvestigator.com/2010/06/11/time-you-enjoy/))

The Cook cooks your tasks and serves them in a better format.
---
It is...

- easy to learn
- text-based
- fast to use
- better than your brain at remembering

---

All you need to do is:

1. [DOWNLOAD](https://github.com/lm-44/ip/releases/tag/A-Jar) NOW!!?!
2. double-click it.
3. realise v0.1 doesn't work  :sob:
4. resort to other means of remembering

It is **NOT FREE**, as it takes up more time than it saves.

What can it do?

- [x] Waste your time

<h1>DISCLAIMER:</h1>
<sup><s>v0.1 doesn't work</s> it still doesn't work</sup>

---

If you are a Java programmer, you can use it to learn what not to do. Here's the main method:
```java
public class Cook {
  public static void main(String[] args) {
          new Cook(new File("data", "tasks.txt")).run();
      }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
renamed Duke.java to Cook.java and added Level-0 functionality

Commit:
added Level-1 functionality, changed font of ascii art and gave credits to original authors

Commit:
added Level-2 functionality and changed text formatting

Commit:
added Level-3 functionality and renamed some variables to better describe its purpose

Commit:
added Level-4 functionality

Commit:
Revert "added Level-4 functionality"

This reverts commit 815e310006a6368b2d35c3df6a14b65b0995bafd.

Commit:
Reapply "added Level-4 functionality"

This reverts commit 39999bc0c7e404c92b3afcb2f91ec71239f08e5d.

Commit:
added A-TextUiTesting functionality

Commit:
added Level-5 functionality

Commit:
added Level-6 functionality

Commit:
correct boolean naming conventions

Commit:
Revert "correct boolean naming conventions"

This reverts commit 0b560a86d454dec26683254ecc45b7c1253b39ab.

Commit:
Correct boolean naming conventions

Commit:
Add startup testing

Commit:
Fix errors, add Level-7 functionality

Commit:
Test branching

Commit:
Merge branch 'branch-Level-7'

Commit:
Add ACTUAL.TXT

Commit:
Correct automated testing error

Commit:
Add Level-8 functionality

Commit:
Test branch-Level-8

Commit:
Merge branch 'branch-Level-8'

Commit:
Add A-MoreOOP functionality

Commit:
Add A-Packages functionality

Commit:
Merge branch 'add-gradle-support' of https://github.com/lm-44/ip

Commit:
Add A-JUnit functionality

Commit:
Create jar file

Commit:
Create jar file

Commit:
Merge branch 'master' of https://github.com/lm-44/ip

Commit:
Add A-JavaDoc functionality

Commit:
Add A-CodingStandard functionality

Commit:
Add Level-9 functionality

Commit:
Test merge conflict

Commit:
Revert "Test merge conflict"

This reverts commit a339fd1f12de6286c785eef580ae0bcd1e6e61c6.

Commit:
Test merge conflict

Commit:
Merge tag 'A-CodingStandard'

Commit:
Merge tag 'A-JavaDoc'

# Conflicts:
#	src/main/java/cook/Cook.java
#	src/main/java/cook/Parser.java
#	src/main/java/cook/TaskList.java

Commit:
Resolve merge conflicts

Commit:
Resolve another merge conflict

Commit:
Add comment for merging Level-9 into master

Commit:
Merge branch 'heads/Level-9' of https://github.com/lm-44/ip

# Conflicts:
#	src/main/java/cook/Cook.java

Commit:
Add A-CheckStyle

Commit:
Merge tag 'A-CheckStyle'

AgentHagu commented on src/main/java/cook/Cook.java:
```diff
@@ -0,0 +1,154 @@
+ */
+public class Cook {
+    private Storage storage;
+    private TaskList tasks;
```
I like how you used plural form for your TaskList object since it's a collection of Tasks, good job!

AgentHagu commented on src/main/java/cook/Cook.java:
```diff
@@ -0,0 +1,154 @@
+                    continue;
+                }
+
+                boolean toMark = command.equals("mark");
```
Perhaps you could rename this boolean `toMark` to use another prefix to better follow the coding standard?

ty-3401 replied:
🫡

AgentHagu commented on src/main/java/cook/Parser.java:
```diff
@@ -0,0 +1,65 @@
+     * @return HashMap of commands and their corresponding inputs.
+     * @throws InvalidCommandException If command is not recognized.
+     */
+    public HashMap<String, String> readInput(String input) throws InvalidCommandException {
```
I like the use of a verb for this method `readInput`, improving the code quality. Good job!

AgentHagu commented on src/main/java/cook/Storage.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Storage class to store file locally.
+ */
+public class Storage {
```
I like the use of a noun for the class, improving the code quality. Good job!

AgentHagu commented on src/main/java/cook/Cook.java:
```diff
@@ -0,0 +1,154 @@
+            } else if (command.equals("list")) {
+                this.ui.say(this.tasks.toString());
+            } else if (command.contains("mark") || command.equals("delete")) {
+                int taskNo;
```
While in general people will understand "no" as "number", perhaps you could use the longer form of the word to avoid possible confusion and improve code quality?

ty-3401 replied:
🫡

AgentHagu commented on src/test/java/cook/ParserTest.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    @Test
+    public void byeTest() {
```
Perhaps the test method names could be improved to better follow the three part format in the coding standard?

ty-3401 replied:
🫡

AgentHagu reviewed (commented):
Overall, good job with following the coding standard and code quality guidelines. You used nouns and verbs for classes/variables and methods respectively.

There are only a few nitpicks to fix.

lm-44 commented on src/test/java/cook/ParserTest.java:
```diff
@@ -0,0 +1,206 @@
+
+import java.util.HashMap;
+
+import org.junit.jupiter.api.Test;
```
I think this import statement should be moved to line 4 (or line 3 to line 8), to maintain the import statement consistency. https://se-education.org/guides/conventions/java/intermediate.html#statements

Commit:
Add GUI, fixed nits by @AgentHagu

Commit:
Merge tag 'Level-10'

Commit:
Add credits

Commit:
Refactor commands

Commit:
Add ByeCommand class

Commit:
Execute function in commands: return String instead of void

The GUI is not able to retrieve messages printed to console in the
CLI.

Returning String allows the GUI to retrieve the messages

ty-3401 committed:
Merge pull request #1 from lm-44/branch-A-CodeQuality

Branch a code quality

Commit:
Fix bugs

ty-3401 committed:
Merge pull request #2 from lm-44/branch-A-CodeQuality

Fix bugs

Commit:
Merge branch 'master' of https://github.com/lm-44/ip

Commit:
Add assert

Commit:
Merge branch 'master' of https://github.com/lm-44/ip into branch-A-Assertions

# Conflicts:
#	src/main/java/cook/Cook.java

ty-3401 committed:
Merge pull request #3 from lm-44/branch-A-Assertions

Add assert

Commit:
Fix load file

Loading a file locally did not work.

Implementing Serializable on Tasks and TaskList
allowed easier saving and loading from a local file

The local file does not need to be human readable, therefore,
this is the simplest way to implement the saving and loading
feature.

Commit:
Add C-DetectDuplicates functionality

Tasks with the same description can be added to the list.

Tasks with the same description may have been entered erroneously
by the user.

rayray39 commented on src/main/java/commands/ToDoCommand.java:
```diff
@@ -0,0 +1,46 @@
+     * Adds ToDo task and saves file.
+     */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
```
For this execute method that is common for ToDoCommand, EventCommand and DeadlineCommand, perhaps you could consider putting it in Command class instead so that you don't have to repeat yourself. But great job on implementing this!

ty-3401 replied:
Ah! The main reason I separated them out was because each command would need to create a specific instance of a different class.
``` java
// Create ToDo instance
new Todo(this.description)
// Create Deadline instance
new Deadline(this.description, this.by)
// Create Event instance
new Event(this.description, this.from, this.to)
```
I don't know of a clean way to combine all of this into a single method in the command class since it only contains the description attribute. Even if I did, the method would be inherited by the other commands which I don't really want to happen since they don't even make use of it.

The other option I thought of was to create another abstract class inheriting from Command, which all task creating commands will inherit from. This will reduce code duplication by containing it within the new abstract class, but I'm unsure of the feasibility of calling a super method within its overridden method in Java. Do let me know if you have other thoughts on this.

Commit:
Fix tests and localised commands within Parser class

Commit:
Add better JavaDocs

Commit:
Change execution to GUI

Commit:
Change execution to GUI for real

Commit:
Fix bugs, add screenshot, change images

Commit:
Add user guide

Commit:
Fix save/load file

Commit:
Remove redundant comment
